### PR TITLE
Adding delay on replication task processing

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
1. wait time before processing batch task per shard
2. wait time between each task

**Why?**
1. Avoid burst load on hosts when fetching tasks for shards
2. Slow down task processing rate

**How did you test it?**

**Potential risks**

### CODE REVIEW

File: service/history/replication/task_processor.go [hunk 1]
```diff
@@ -121,16 +123,19 @@ func NewTaskProcessor(
...
+		sourceCluster:     taskFetcher.GetSourceCluster(),
+		status:            common.DaemonStatusInitialized,
+		shard:             shard,
+		historyEngine:     historyEngine,
+		historySerializer: persistence.NewPayloadSerializer(),
+		config:            config,
+		metricsClient:     metricsClient,
+		logger:            shard.GetLogger(),
+		taskExecutor:      taskExecutor,
+		rateLimiter: quotas.NewDynamicRateLimiter(func() float64 {
```
Comment (yycptt) [inline service/history/replication/task_processor.go:136]: Is this rateLimiter used anywhere?

File: service/history/replication/task_processor.go [hunk 2]
```diff
@@ -121,16 +123,19 @@ func NewTaskProcessor(
...
+			return config.ReplicationTaskProcessorQPS(shardID)
```
Comment (yycptt) [inline service/history/replication/task_processor.go:137]: I think in longer term we should enforce QPS on host level?

Comment (yux0) [reply service/history/replication/task_processor.go]: Added the host level rate limiter

File: service/history/replication/task_processor.go (hunk 1 above)
Comment (yux0) [reply service/history/replication/task_processor.go]: Updated to use host level rate limiter

File: service/history/replication/task_processor.go [hunk 3]
```diff
@@ -285,12 +298,15 @@ func (p *taskProcessorImpl) processResponse(response *r.ReplicationMessages) {
...
 	batchRequestStartTime := time.Now()
+	ctx := context.Background()
 	for _, replicationTask := range response.ReplicationTasks {
+		p.rateLimiter.Wait(ctx)
 		err := p.processSingleTask(replicationTask)
 		if err != nil {
 			// Processor is shutdown. Exit without updating the checkpoint.
 			return
 		}
+		time.Sleep(p.config.ReplicationTaskProcessorTaskProcessingWait(p.shard.GetShardID()))
```
Comment (yycptt) [inline service/history/replication/task_processor.go:309]: Synced offline. Let's change this to a shard level rate limiter.

Comment (yux0) [reply service/history/replication/task_processor.go:309]: Updated

File: service/history/replication/task_fetcher.go [hunk 4]
```diff
@@ -167,8 +170,11 @@ func newReplicationTaskFetcher(
 		remotePeer:     sourceFrontend,
 		currentCluster: currentCluster,
 		sourceCluster:  sourceCluster,
-		requestChan:    make(chan *request, requestChanBufferSize),
-		done:           make(chan struct{}),
+		rateLimiter: quotas.NewDynamicRateLimiter(func() float64 {
```
[the remark this answers is not in the document]
Comment (yux0) [reply service/history/replication/task_fetcher.go:173]: Synced offline. Leave it in the fetcher at this point

File: service/history/config/config.go [hunk 5]
```diff
@@ -430,20 +434,24 @@ func New(dc *dynamicconfig.Collection, numberOfShards int, storeType string, isA
...
+		ReplicationTaskProcessorErrorRetryWait:             dc.GetDurationPropertyFilteredByShardID(dynamicconfig.ReplicationTaskProcessorErrorRetryWait, 50*time.Millisecond),
+		ReplicationTaskProcessorErrorRetryMaxAttempts:      dc.GetIntPropertyFilteredByShardID(dynamicconfig.ReplicationTaskProcessorErrorRetryMaxAttempts, 5),
+		ReplicationTaskProcessorNoTaskRetryWait:            dc.GetDurationPropertyFilteredByShardID(dynamicconfig.ReplicationTaskProcessorNoTaskInitialWait, 2*time.Second),
+		ReplicationTaskProcessorCleanupInterval:            dc.GetDurationPropertyFilteredByShardID(dynamicconfig.ReplicationTaskProcessorCleanupInterval, 1*time.Minute),
+		ReplicationTaskProcessorCleanupJitterCoefficient:   dc.GetFloat64PropertyFilteredByShardID(dynamicconfig.ReplicationTaskProcessorCleanupJitterCoefficient, 0.15),
+		ReplicationTaskProcessorReadHistoryBatchSize:       dc.GetIntProperty(dynamicconfig.ReplicationTaskProcessorReadHistoryBatchSize, 5),
+		ReplicationTaskProcessorStartWait:                  dc.GetDurationPropertyFilteredByShardID(dynamicconfig.ReplicationTaskProcessorStartWait, 3*time.Second),
```
Comment (meiliang86) [inline service/history/config/config.go:448]: I feel the max wait can be bigger. Set default to 5s?

File: service/history/config/config.go [hunk 6]
```diff
@@ -430,20 +434,24 @@ func New(dc *dynamicconfig.Collection, numberOfShards int, storeType string, isA
...
+		ReplicationTaskProcessorStartWaitJitterCoefficient: dc.GetFloat64PropertyFilteredByShardID(dynamicconfig.ReplicationTaskProcessorStartWaitJitterCoefficient, 0.8),
```
Comment (meiliang86) [inline service/history/config/config.go:449]: Make it 1?

File: service/history/config/config.go (hunk 5 above)
Comment (yux0) [reply service/history/config/config.go]: updated

File: service/history/config/config.go (hunk 6 above)
Comment (yux0) [reply service/history/config/config.go]: need to be less than 1. updated it to 0.9

File: service/history/replication/task_processor.go [hunk 7]
```diff
@@ -78,6 +79,8 @@ type (
 		metricsClient     metrics.Client
 		logger            log.Logger
 		taskExecutor      TaskExecutor
+		hostRateLimiter   *quotas.DynamicRateLimiter
+		shardRateLimiter  *quotas.DynamicRateLimiter
```
[the remark this answers is not in the document]
Comment (yux0) [reply service/history/replication/task_processor.go:83]: done

**Pull request merged**